csharp/src/Ice/Connection.cs

```diff
@@ -68,145 +65,111 @@ public enum ConnectionClose
         GracefullyWithWait
     }
 
-    public sealed class Connection : IceInternal.EventHandler, ICancellationHandler
+    public sealed class Connection : ICancellationHandler
     {
-        public interface IStartCallback
-        {
-            void ConnectionStartCompleted(Connection connection);
-            void ConnectionStartFailed(Connection connection, System.Exception ex);
-        }
-
-        private class TimeoutCallback : ITimerTask
-        {
-            public TimeoutCallback(Connection connection) => _connection = connection;
-
-            public void RunTimerTask() => _connection.TimedOut();
-
-            private readonly Connection _connection;
-        }
-
-        public void Start(IStartCallback? callback)
+        /// <summary>Get or set the object adapter that dispatches requests received over this connection.
+        /// A client can invoke an operation on a server using a proxy, and then set an object adapter for the
+        /// outgoing connection used by the proxy in order to receive callbacks. This is useful if the server
+        /// cannot establish a connection back to the client, for example because of firewalls.</summary>
+        /// <returns>The object adapter that dispatches requests for the connection, or null if no adapter is set.
+        /// </returns>
+        public ObjectAdapter? Adapter
         {
-            try
+            get
             {
-                lock (this)
+                lock (_mutex)
```

Comment:
Added:
```
        public ObjectAdapter? Adapter
        {
            // We don't use a volatile for _adapter to avoid extra-memory barriers when accessing _adapter with
            // the mutex locked.
```